### PR TITLE
Generate schema / references for enum action parameters #358

### PR DESCRIPTION
**Pull request by onionhammer**
This basically changes the behavior of `DescribeAllEnumsAsStrings` for Enum types to generate a `Ref` for the parameter rather than generate the supported values inline.

Addresses #358

### CODE REVIEW

File: src/Swashbuckle.AspNetCore.SwaggerGen/Generator/SwaggerGenerator.cs [hunk 1]
```diff
@@ -192,8 +192,13 @@ private Operation CreateOperation(ApiDescription apiDescription, ISchemaRegistry
             if (schema == null)
                 nonBodyParam.Type = "string";
             else
+            {
                 nonBodyParam.PopulateFrom(schema);
 
+                if (schema.Ref != null)
+                    nonBodyParam.Ref = schema.Ref;
```
Comment (domaindrivendev) [inline src/Swashbuckle.AspNetCore.SwaggerGen/Generator/SwaggerGenerator.cs]: A very minor change, but could you move the following line into `SchemaExtension.PopulateFrom`:

```csharp
nonBodyParam.Ref = schema.Ref;
```

No need the for the `if` condition either

**Comment by domaindrivendev** [comment]
I'm pretty much ready to merge this - just one minor change proposed.

**Comment by domaindrivendev** [comment]
Also ... on second thoughts, I wonder if we should do this for all enums, regardless of the `DescribeAllEnumsAsStrings` setting. I'd prefer to keep the behavior consistent regardless of how client generators deal with it.

**Comment by onionhammer** [comment]
@domaindrivendev I'll takey a look at changing that tomorrow

Regarding `DescribeAllEnumsAsStrings` I'm fine with that if you are

**Comment by onionhammer** [comment]
Okay; I moved the code to assign the `Ref`, but it involves some unboxing and reboxing the partial schema since `PartialSchema` doesnt have the `Ref` property.

I can eliminate the `DescribeAllEnumsAsStrings` check as well if you like; but doesnt it have the potential to break people's code who upgrade swashbuckle and regenerate their code? I suppose it would force them to use a value thats closer to what they would want to see anyways though.

**Comment by onionhammer** [comment]
@domaindrivendev Want me to update this PR to always generate the enum names? Its a small change, but it could have implications I dont know of right now - even when I get the tests passing again

**Comment by onionhammer** [comment]
@domaindrivendev Okay.. just in case you want it this way I now generate enums as references regardless of the config

**Comment by onionhammer** [comment]
@domaindrivendev I would actually like to make use of this some day, as it's the defacto-standard for aspnet core (even used in Microsoft's own docs), you requested assistance in the referenced issue:

> This should be a relatively straightforward change. Anyone on this thread interest in submitting a PR? This is FOSS afterall :)

And I came through with a PR, but I guess I'll have to write a swagger document by-hand or continue using WebApiProxy for now, as this is such a glaring limitation that would affect so much of my existing (generated) API.

**Comment by domaindrivendev** [comment]
@onionhammer - I guess this isn't as straight-forward as I first thought. After seeing your message above, I made some time to merge the changes, only to discover that they cause the generated Swagger to be invalid.

In a nutshell, Swagger 2.0 does NOT support the `$ref` field on [Parameter Objects](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#parameterObject) and therefore, a shared enum definition is NOT possible for non-body parameters. Full details are discussed [here in the Swagger/OpenAPI repo](https://github.com/OAI/OpenAPI-Specification/issues/564).

So, we're snookered until SB supports v3 of the Swagger spec. This is a much heavier lift. In the meantime, I can merge a partial version of this PR that supports shared enum definitions within body parameters and response objects ONLY.

On a separate note, regarding your concern around backwards compatibility, I think we should put this behavior behind a specific flag that you need to opt into (e.g. `CreateReferencesForEnums`). This way, no-one is going to be surprised by the new behavior. With the next major release, where breaking changes are more acceptable, I might remove the flag and default to the new behavior.

I'll go ahead and make the adjustments. Please let me what you think - I'm assuming you're not going to be too happy about this, but an implementation that breaks the spec (even if the UI honors it) just isn't on the table.

**Comment by onionhammer** [comment]
@domaindrivendev thanks for looking into this, it's a shame that this breaks the spec, but it's better you found out now than later

Looking forward to OpenApi/v3

**Comment by domaindrivendev** [comment]
NOTE: I applied the changes I described above and merged. Because, I applied them to the merge commit, they're not going to show up in the `Files changed` tab for this PR. To see the final set of changes that got merged refer to the merge commit - https://github.com/domaindrivendev/Swashbuckle.AspNetCore/commit/5abcea4f59a9061870581d99dafbca3524e6b2fa.

In a nutshell, you can now opt into referenced definitions for enums that occur within body parameters OR response schemas with the following config option. Unfortunately, inline enum definitions will still be used for non-body parameters due to the Swagger 2.0 limitations described above.

**Comment by yahorsi** [comment]
Hi Guys,

We recently had to move to the .NET Core 3.1 as 2.2. support ended and so, had to move to the latest Swashbuckle. Our models code-gen does not support Open Api 3.0 so, we have to use: UseSwagger(x=>x.SerializeAsV2 = true);

Unfortunately, in the generated swagger spec we see enums as references that is not supported by our code-gen as well.

Is there any way to turn old 4.0.1 behavior?

Thanks for the answers!

**Pull request merged**